packages/cli-kit/src/plugins.ts

```diff
@@ -70,4 +88,14 @@ export type FanoutHookFunction<
 > = (
   this: Interfaces.Hook.Context,
   options: TPluginMap[TEvent]['options'] & {config: Interfaces.Config},
-) => Promise<Partial<PluginReturnsForHook<TEvent, TPluginName, TPluginMap>>>
+) => Promise<PluginReturnsForHook<TEvent, TPluginName, TPluginMap>>
+
+export const tunnel = {
+  defineProvider: (input: {name: string}) => () => input,
+  startTunnel: (options: {provider: string; action: (port: number) => Promise<string | undefined>}) => {
+    return (inputs: {port: number; provider: string}) => {
+      if (inputs.provider !== options.provider) return undefined
+      return options.action(inputs.port)
+    }
+  },
```

Comment:
By having this middleware we can control which plugins are actually called.

Each plugin defines a provider and an action to be executed, then when the hook is called the action will be run only if the defined provider is the same as the intended provider (sent via runHook).

The types between this functions and `HookReturnsPerPlugin` are not the same, I'd need to find a way to tie them.